### ISSUE TITLE
Pass a Func<string> to lazily evaluate an exception message

### ISSUE THREAD

**Issue by pflugs30** [nunit/nunit#144]
I find several cases in my tests where I only want to build a message when the Assertion fails.  Consider the following snippet:

// Assert
var comparer = new MyEqualityComparer();
Assert.That(actual, Is.EqualTo(expected).Using(comparer));

I have several cases where I use an IEqualityComparer<T> to simplify multiple Assertion statements into a reusable EqualityComparer.  It's great, until the the verification fails, and I have to manually step through the test comparing properties manually.

To help automate this task, I built a reusable helper class to not only compare two classes but enumerate the differences between their properties.  Then, the differences are rendered to a string, and I can use this string as the failure message in the Assertion.

Ideally, this message would not need to be created unless the verification fails.  However, none of the Assert methods allow for passing a function as a parameter which would lazily evaluate the failure message.  Here's the syntax I'd like to use:

// Assert
var comparer = new MyEqualityComparer();
Func<string> getFailureMessage = () => comparer.WriteDifferencesToString(expected, actual);
Assert.That(actual, Is.EqualTo(expected).Using(comparer), getFailureMessage);

Using this syntax, client test methods could lazily evaluate the failure messages using any valid Function.  I think this would really help users clean up their tests and extend the capabilities of the NUnit framework.

I'm open to discussion on this idea as well as direction on how to help make it a possibility.  Thank you for your consideration.

**Comment by CharliePoole** [comment]
I can't believe this has been hanging about for two years! It's a nice idea although not top priority. Maybe taking it out of the ideas category and marking it as a feature will bring it more notice.

**Comment by jnm2** [comment]
+1 I looked for this feature and missed it badly today.

**Comment by CharliePoole** [comment]
@ChrisMaddock Are you interested in working on this one?

**Comment by pflugs30** [comment]
I originally made this request, so I feel a sense of responsibility to make it happen.  I won't be able to work on it for at least two weeks, however.  If someone wants it bad enough before then, have at it.  Otherwise, I'll take a stab at it in a few weeks.

**Comment by ChrisMaddock** [comment]
@CharliePoole - happy to leave it to @jnm2 or @pflugs30 if they're interested. Good idea though!

**Comment by pflugs30** [comment]
@CharliePoole: Thank you for the invite, sir.  I'll take you up on that offer.  Due to some personal commitments, I can't dedicate a lot of time for about two weeks, but I will work on this feature As soon as possible.

**Comment by pflugs30** [comment]
@CharliePoole: An update for you, sir...  I have forked the repo and will begin working on this issue this weekend as time permits.

**Comment by CharliePoole** [comment]
Good to know. Thanks!

**Comment by pflugs30** [comment]
Update...  I've added two main overloads to the `Assert.That<TActual>` method, one that accepts `Func<ConstraintResult, string>` and another to accept a `Func<string>`instead of nothing or a string format and parameter arguments.  The former is intended to pass the result of the `IResolveConstraint.ApplyTo(actual)` call back to the client so it can be used in formatting the custom message.  The latter allows the client to simply pass a function that doesn't need the result.

The change should give us much more flexibility in the Exception messages we can build without having to create our own classes.

I've been working through the various Assert methods adding overloads to most operations.   I have yet to start updating the unit tests.  I will likely be to a point for review within a week or so, as time permits.

**Comment by CharliePoole** [comment]
Would you post some examples of how this might be used by users? I'm a little concerned with having users deal with ConstraintResult, which is an implementation class. Seeing the examples might inspire us with some alternative suggestions for signatures.

Which Assert methods were you aiming to change? Your "working through the assert methods" comment has me slightly worried. :-) My thought was that this issue was about Assert.That and Assume.That.

**Comment by pflugs30** [comment]
Hi @CharliePoole.  Sure thing.  I want to make sure I'm going the right way, too

Back to my original post, I used the example where I wanted to use a custom comparer when ensuring that two objects are equal.  Then if the equality comparison fails, we build a custom error message to simplify the process of determining WHY they are not equal.  Here's a pseudo-code example:

```
var comparer = new MyEqualityComparer();
Func getFailureMessage = () => comparer.WriteDifferencesToString(expected, actual);
Assert.That(actual, Is.EqualTo(expected).Using(comparer), getFailureMessage);
```

This method to write differences may use reflection, and thus it could be fairly slow.  We don't want to do it unless the Assertion fails.

To your question about why they could use the `ConstraintResult`, that actually came from the refactoring I needed to do to the Assume/Assert.That methods to allow me to call this function parameter.  Here's my refactored Assert.That method:

```
public static void That<TActual>(TActual actual, IResolveConstraint expression, Func<ConstraintResult, string> getExceptionMessage)
        {
            var constraint = expression.Resolve();
            IncrementAssertCount();
            var result = constraint.ApplyTo(actual);
            if (!result.IsSuccess)
            {
                throw new AssertionException(getExceptionMessage(result));
            }
        }
```

Here's the case where the user doesn't pass their own function:
```
        public static void That<TActual>(TActual actual, IResolveConstraint expression, string message, params object[] args)
        {
            Assert.That(actual, expression, BuildDefaultExceptionMessageFunc(message, args));
        }
```

And...
```
        public static Func<ConstraintResult, string> BuildDefaultExceptionMessageFunc(string message, object[] args)
        {
            return result =>
            {
                MessageWriter writer = new TextMessageWriter(message, args);
                result.WriteMessageTo(writer);
                var exceptionMessage = writer.ToString();
                return exceptionMessage;
            };
        }

```

If we don't want the user interacting with the ConstraintResult, it would be fairly simply to prevent this.  I have a helper method to allow the user to pass in a `Func<string>` instead of a `Func<ConstraintResult, string>`, so we could simply mark the ConstraintResult overload as non-public.


Now, finally, to your comment about which Assert methods I am changing, I was adding overloads to the various Assertion methods to allow a user to pass in either the message/args combination for string.Format OR the Func<string>.  Here's an example from Assert.Equality (partially elided for space:):

```
public static void AreEqual(object expected, object actual, string message, params object[] args) {}
public static void AreEqual(object expected, object actual, Func<ConstraintResult, string> getExceptionMessage) {}
public static void AreEqual(object expected, object actual, Func<string> getExceptionMessage) {}
public static void AreEqual(object expected, object actual) {}

```

These methods simply forward into Assert.That.

Again, I'm working in my own forked repository.  I've got what I think is a reviewable copy.  Do these answers help set the stage?

**Comment by pflugs30** [comment]
@ChrisMaddock or @jnm2, can you please take a look at the first code block with the pseudo code in my above example to see if this is something like you were trying to do?

**Comment by pflugs30** [comment]
Also, for convenience, here's my fork:
https://github.com/pflugs30/nunit

**Comment by CharliePoole** [comment]
Leaving on a trip just now. I will try to reply to you tonight.

**Comment by jnm2** [comment]
`Func<string>` would work for me as in your first code block. I wouldn't be using a comparer, so this is what I would do:

```c#`
Assert.That(validationErrors, Is.Empty, () => string.Join("\r\n", validationErrors));
```

**Comment by pflugs30** [comment]
Thanks, @jnm2.  That's exactly what I was going for.

I realize now that @CharliePoole was likely meaning that I should NOT have added the overloads to the "outdated" [Classic Model](https://github.com/nunit/docs/wiki/Classic-Model)...  That would have saved me a lot of work... ;-)

**Comment by pflugs30** [comment]
@CharliePoole:  I know you're travelling, so there's no hurry here.  I wanted to mention that I looked through the documentation with a new eye now based on your comment, and I fully expect I will need to revert the changes I made to the Assert methods **other than** the ones made to `Assert.That` and `Assume.That`.  I will wait to hear back from you.

**Comment by pflugs30** [comment]
Hi @CharliePoole.  I will have some time to work on the code this weekend.  Did you have a chance to look at the comments above or the modifications in my fork?  If not, I completely understand.  I think I'll be needing to pull quite a few new overloads back that I added to the Classic Model code...  That's all.

**Comment by ChrisMaddock** [comment]
@pflugs30 - Sorry, been away! Sample looks great to me. ☺

Regarding the classic model, I can't speak for @CharliePoole, but i'd say, I you've already done the work, submit it anyway - from my limited experience, the classic model isn't religiously kept up to date with new features, but they're welcomed as submissions.

**Comment by CharliePoole** [comment]
I'm not in favor of adding this feature to the classic asserts but of course we can discuss it.

@ChrisMaddock I think you have been around for a few exceptions to the rule, but we have generally refused to modify the classic (legacy) model. Both terms (classic and legacy) are intended to give people that impression but unfortunately words sometimes have different connotations for different people. Recently we added Assert.Zero and Assert.NotZero but only after extensive argument.

In terms of contributions, it's always a good idea to put different features into separate PRs. If a PR has two different features and we reject one but like the other, that causes the submitter a bit of work. In this case, though, the original discussion was ambiguous. I suggest that @pflugs30 should leave the code as it is until other folks chime in. @rprouse ? @oznetmaster ? What do you think?

**Comment by pflugs30** [comment]
Hi Team, thanks for commenting.

I would also prefer NOT to include changes to the Classic (Legacy) model, now that I better the way to do things.  As such, I created a branch from my original fork and began again, copying the changes to Assert, Assert.That, and Assume that I needed to keep.  I also took the next step of adding tests for these modifications.  The Assertion Tests all pass (existing and new), though I see that some of the tests in the solution do NOT pass.  I think these failing tests were failing previously.

I think I have a version that is ready for a pull request (or, at minimum, peer review).  Any advice on the way to proceed next?  Go ahead with the PR?  Thanks.

**Comment by CharliePoole** [comment]
A PR is the way to go, if it's not actually ready for merge in your own view, just add that comment at the top of the description. We'll make comments about any changes we think are needed and ask questions. Two committers will have to OK it before it gets merged.

**Comment by pflugs30** [comment]
Sounds good.  I will attempt to do so.  I'm really unfamiliar with how to
use GitHub, so I'll admit I'm struggling with the merging, branching, and
pull request processes.  I'm much stronger in SVN... :-)

Thanks,
Matthew
